### DISCUSSION TITLE
[Idea] Set up different subdomain for backoffice

### DISCUSSION THREAD

**Discussion by cedricfontaine** [PrestaShop/PrestaShop#32272] [Ideas]
For security reason, we would like to serve the back office on a different subdomain than the front office.
In Prestashop 1.6 and previous, it was possible to do it. It's not anymore the case.
We would like an option to be able to separate the front and back office subdomain url.

**Comment by khouloudbelguith** [comment]
Hi @cedricfontaine,

Thanks for your report.
@PrestaShop/prestashop-core-developers what do you think?

Thanks!

**Comment by PierreRambaud** [comment]
Can you tell us more about why it's impossible to separate front & back? I just remove my admin directory, change my shop domain and my front's still working well when my back is on another domain. My back's working well too.
Running the latest develop branch.

**Comment by cedricfontaine** [comment]
In 1.6, it was possible to access back office from an url different from the shop_url, it's not the case anymore in 1.7, you're always redirected to the shop_url.
We would like to be able either to disable this redirect or specify an url for the back office.

**Comment by PierreRambaud** [comment]
Hey thanks for teaching me something!

If you add in your config.inc.php (for example,)
```
define('_PS_BASE_URL_', 'http://ps-test.localhost');
```
Your website'll have a different url.

I'm able to run a front & a back from different domains / subdomains.
Also working with `_PS_BASE_URL_SSL_`

**Comment by khouloudbelguith** [comment]
Hi @cedricfontaine,

Any news about your suggestion?
Thanks!

**Comment by cedricfontaine** [comment]
Hello,

We plan to test it in the new few days

On Wed, Apr 24, 2019 at 4:43 AM khouloudbelguith <notifications@github.com>
wrote:

> Hi @cedricfontaine <https://github.com/cedricfontaine>,
>
> Any news about your suggestion?
> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PrestaShop/PrestaShop/issues/13016#issuecomment-486127970>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYIMUEV26RU52KODZKUTRTPSAMULANCNFSM4HAPQ36A>
> .
>


-- 

*Cédric Fontaine*
Équipe E-commerce Team
Décathlon Canada
p: +1 514-967-4441
w: www.decathlon.ca  e: cedric.fontaine@decathlon.com

**Comment by khouloudbelguith** [comment]
@cedricfontaine, thanks!
Waiting for your feedback.

**Comment by jolelievre** [comment]
@cedricfontaine I'm not sure if this will work as I don't know exactly how this redirection is managed
But maybe you can use the multistore feature to add your second domain and it will allow you to use it (be careful about the multi shop context you are editing though with the second domain)

**Comment by centoasa** [comment]
Also for me: how can I move my admin from https://www.disney.com/admin to--> https://admin.disney.com ?
Thank you

**Comment by khouloudbelguith** [comment]
Hi @centoasa,

You can use the solution of @PierreRambaud https://github.com/PrestaShop/PrestaShop/issues/13016#issuecomment-475702896
You need to add in your config.inc.php 

`define('_PS_BASE_URL_', 'https://admin.disney.com');`

Thanks!

**Comment by centoasa** [comment]
@khouloudbelguith but this solution will change my url shop, or only admin?
Now my admin have this configuration-->https://nimb.ws/2aqpIB
and also I've setup my dns into A admin.100asa.it (ip my server)
but if I try to goes into url admin.100asa.it, put into the defualt shop url www.100asa.it

**Comment by khouloudbelguith** [comment]
@cedricfontaine, in your first screenshot, to access to FO => you need to naviagate to this link https://www.100asa.it/
To access to your BO => you need to navigate to this link https://www.100asa.it/admin_folder_name
In your case, you need to change this https://www.100asa.it/admin_folder_name to this https:/www.admin.100asa.it?
Thanks!

**Comment by centoasa** [comment]
@khouloudbelguith yes it's true! I would to access to my admin via https://admin.100asa.it to bypass 100 seconds cloudflare timeout scripts

**Comment by khouloudbelguith** [comment]
@centoasa,the solution is proposed here https://github.com/PrestaShop/PrestaShop/issues/13016#issuecomment-475702896
Your website'll have different urls.
You will be able to run a front & a back from different domains / subdomains.
Thanks!

**Comment by centoasa** [comment]
@khouloudbelguith I've to rebuild .htaccess or change my default url on my backoffice? (please take a look at https://nimb.ws/k8LhvF )
because for now the #13016 solution for me doesn't works!

**Comment by khouloudbelguith** [comment]
@centoasa, you need to edit the file `config.inc.php` manuaally which exists in this path /Projet_Folder/config/config.inc.php as described here https://github.com/PrestaShop/PrestaShop/issues/13016#issuecomment-475702896
Thanks!

**Comment by centoasa** [comment]
[config.inc.php.zip](https://github.com/PrestaShop/PrestaShop/files/3131732/config.inc.php.zip)
here attached my config (please browse to https://back100asa.100asa.it)

**Comment by khouloudbelguith** [comment]
@centoasa, so with this link https://back100asa.100asa.it you will be redirected to the FO.
ping @PierreRambaud what do you think?
Thanks!

**Comment by PierreRambaud** [comment]
If you want to have a different admin url and front url, you need to have two directories for now, there is currently no distinction between both in http/https configuration.

**Comment by centoasa** [comment]
@PierreRambaud sorry I don't understand

**Comment by PierreRambaud** [comment]
You need to split directories:

`/var/www/front` => `www.website.com`
`/var/www/back` => `admin.website.com`

You can't setup `/var/www/website` for `www.website.com` & `admin.website.com`.
Because of `.htaccess` which is automatically generated, you need to write your own server configuration.
Because of redirection, without the change of PS_BASE_URL, you'll be redirected to the front url.
Maybe you can use a reverse proxy?

**Comment by centoasa** [comment]
@PierreRambaud  my request only to prevent Cloudflare 100 seconds timeout for some my admin scripts; now I've a subdomain named cron so I invoke cron.100asa.it to launch prestashop cron module (many module works well, some goes into timeout)
as you can see into my attached config file, I was made PS_BASE_URL adding, but always redirection on front.

**Comment by PierreRambaud** [comment]
Yep, many cron tasks are running without bootstrapping this is why you can call these scripts.
As I said, with two directories, I'm able to use different domain for my back and my front.

**Comment by centoasa** [comment]
@PierreRambaud by my plesk panel, I've created a subdomain named back100asa and copied the full admin directory (100asa.it/admin) into it https://nimb.ws/cbdy3Q
Launch back100asa.100asa.it I've error 500

**Comment by PierreRambaud** [comment]
We need more information than "500", this means, it's a Internal Server Error.

**Comment by centoasa** [comment]
as you can see from my screenshot, into subdomain back... I've copied all files and folder of prestashop admin. And on /config/config.inc.php I've add the line
define('_PS_BASE_URL_', 'https://back100asa.100asa.it');
 That's all.
here the apache error:
`

AH01071: Got error 'PHP message: PHP Warning: require(/var/www/vhosts/100asa.it/back100asa.100asa.it/../config/config.inc.php): failed to open stream: No such file or directory in /var/www/vhosts/100asa.it/back100asa.100asa.it/index.php on line 42\nPHP message: PHP Warning: require(/var/www/vhosts/100asa.it/back100asa.100asa.it/../config/config.inc.php): failed to open stream: No such file or directory in /var/www/vhosts/100asa.it/back100asa.100asa.it/index.php on line 42\nPHP message: PHP Fatal error: require(): Failed opening required '/var/www/vhosts/100asa.it/back100asa.100asa.it/../config/config.inc.php' (include_path='.:/opt/plesk/php/7.2/share/pear') in /var/www/vhosts/100asa.it/back100asa.100asa.it/index.php on line 42\n'
--


`

**Comment by PierreRambaud** [comment]
This file doesn't exists `/var/www/vhosts/100asa.it/back100asa.100asa.it/../config/config.inc.php` or is not available for reading.

**Comment by centoasa** [comment]
there was a misunderstood: on the folder /var/www/vhosts/100asa.it/back100asa.100asa.it I've only copied all the admin files and folders
the config/config.inc.php is on the root prestashop directory

**Comment by PierreRambaud** [comment]
You need to copy all files, add symlink for shared files between front and back, like generated images, or .htaccess for the front, it's maybe too complicated and you need to have a serious system administrator / developer to do this kind of job.

**Comment by centoasa** [comment]
I assumed ...

**Comment by PierreRambaud** [comment]
But I think we need to create an issue to be able to use the front & the back with the same folder on different domains :)

**Comment by centoasa** [comment]
@PierreRambaud great idea!!!

**Comment by khouloudbelguith** [comment]
ping @PrestaShop/prestashop-core-developers what do you think of using the front & the back with the same folder on different domains as suggested by @PierreRambaud?

Thanks!

**Comment by agnesdennetiere** [comment]
Hello,

any news on this point?

Thanks!

**Comment by PierreRambaud** [comment]
@agnesdennetiere Nothing planned yet. But full free to help us and contribute :pray:

**Comment by Bettelstab** [comment]
I'm trying to do this like @PierreRambaud proposed. As a first step, I just try to change the Backoffice URL on a test instance. So I changed the base URI as described via config file.

I still get redirected to the former URL which is still present in the database (PS_SHOP_DOMAIN / PS_SHOP_DOMAIN_SSL in config table plus table *_SHOP_URL). Is there a way to stop that redirect? Could someone point me to the line where that happens?

PS: I'm on 1.7.5.1.

**Comment by PierreRambaud** [comment]
We should add an option in the backoffice to configure the admin URL and front URL, it's the easiest and best way to do it.

**Comment by jocel1** [comment]
You can put at this end of Shop::setUrl():

 ```      
        $this->domain = $row['domain'];
        $this->domain_ssl = $row['domain_ssl'];

        if (defined('_PS_ADMIN_DIR_') && !empty($_SERVER['HTTP_HOST'])) {
            $this->domain_ssl = $_SERVER['HTTP_HOST'];
            $this->domain = $_SERVER['HTTP_HOST'];
        }

        return true;
```

with this change it will override the fetch from the DB with the current domain only if we are in the admin dir.

As a bonus just to be sure, you can also create a defines_custom.inc.php in the config dir with:
```
<?php
if (defined('_PS_ADMIN_DIR_') && !empty($_SERVER['HTTP_HOST'])) {
    define('_PS_BASE_URL_SSL_', 'https://'.$_SERVER['HTTP_HOST']);
    define('_PS_BASE_URL_', 'http://'.$_SERVER['HTTP_HOST']);
}
```

**Comment by DeFault-1** [comment]
> I'm trying to do this like @PierreRambaud proposed. As a first step, I just try to change the Backoffice URL on a test instance. So I changed the base URI as described via config file.
> 
> I still get redirected to the former URL which is still present in the database (PS_SHOP_DOMAIN / PS_SHOP_DOMAIN_SSL in config table plus table *_SHOP_URL). Is there a way to stop that redirect? Could someone point me to the line where that happens?
> 
> PS: I'm on 1.7.5.1.

Probably this redirect is being forced by the PHP code in **AdminController.php** class or somewhere else.

I am blaming it on **Configuration::get('PS_SHOP_DEFAULT')**

Ideally there should be two separate Configuration variables:

- PS_FRONT_DEFAULT
- PS_BACK_DEFAULT

**Comment by DeFault-1** [comment]
> We should add an option in the backoffice to configure the admin URL and front URL, it's the easiest and best way to do it.

I do absolutely support this very important improvement. Prestashop's Back Office is notoriously sluggish in version 1.7 and very often you have to wait for more than 2 minutes for the page to be ready. Unfortunately Cloudflare will not permit you to wait for more than 100 seconds, that is a big problem.

**Comment by PierreRambaud** [comment]
> > We should add an option in the backoffice to configure the admin URL and front URL, it's the easiest and best way to do it.
> 
> I do absolutely support this very important improvement. Prestashop's Back Office is notoriously sluggish in version 1.7 and very often you have to wait for more than 2 minutes for the page to be ready. Unfortunately Cloudflare will not permit you to wait for more than 100 seconds, that is a big problem.

This is an important improvement, I confirm :+1: 
I plan to work on it in few weeks :)

**Reply by BoutiquesInternet** [reply]
any news ?

**Comment by gyan111** [comment]
Very productive discussion. 

I also want to achieve the same. 

Now:
front: example.com
Back: example.com/admin

Requirement is.
front: example.com
Back: admin.example.com

I have created a subdomain and pointed it to the to the admin folder.
 
The back office html loads but js and css are not loading.

![Opera Snapshot_2021-07-13_090316_admin cairosales com](https://user-images.githubusercontent.com/4457406/125386628-3a82b980-e3ba-11eb-8399-00ccc3d7ebf2.png)

I am getting issues like the "Refused to apply stylesheet because the mimetype not supported". See screenshot

![Untitled](https://user-images.githubusercontent.com/4457406/125386705-5be3a580-e3ba-11eb-92c3-fdd85fdba423.jpg)


I have applied all suggestions discussed above but none of them is working.

Please suggest something if possible.

I am using Prestashop 1.6

**Comment by PierreRambaud** [comment]
> Very productive discussion.
> 
> I also want to achieve the same.
> 
> Now:
> front: example.com
> Back: example.com/admin
> 
> Requirement is.
> front: example.com
> Back: admin.example.com
> 
> I have created a subdomain and pointed it to the to the admin folder.
> 
> The back office html loads but js and css are not loading.
> 
> ![Opera Snapshot_2021-07-13_090316_admin cairosales com](https://user-images.githubusercontent.com/4457406/125386628-3a82b980-e3ba-11eb-8399-00ccc3d7ebf2.png)
> 
> I am getting issues like the "Refused to apply stylesheet because the mimetype not supported". See screenshot
> 
> ![Untitled](https://user-images.githubusercontent.com/4457406/125386705-5be3a580-e3ba-11eb-92c3-fdd85fdba423.jpg)
> 
> I have applied all suggestions discussed above but none of them is working.
> 
> Please suggest something if possible.
> 
> I am using Prestashop 1.6

PrestaShop 1.6 is not maintained anymore, we can't do anything related to this version :/

**Comment by gyan111** [comment]
Ok! No problem.

**Comment by AshleyButcher** [comment]
For anyone who might be interested, for our 8.2.x system (a multistore environment), we locked down the public facing shop, and have a dedicated secure admin vhost. We preferred this to locking down the /adminXXXXXXX location for a variety of reasons, including those already listed in this discussion, but mostly because we use mTLS (client-side certificates) and wanted to avoid TLS renegotiation issues, but that's another story. We also have a dedicated vhost for API access, but that's also another story.

This is how we've implemented this, but of course YMMV.

---

#### 1. Lock down _public vhosts_

Our _public vhosts_ are locked down with this config snippet (this is for Apache - the first location is perhaps redundant, but I'm paranoid):

```apacheconf
        <LocationMatch "^/(admin|api|app|bin|classes|config|controllers|docs|translations|vendor)">
                Require all denied
        </LocationMatch>

        <Directory /srv/www/shop.gassycat.be/htdocs>
                AllowOverride All
                Require all denied
        </Directory>
        
        <Directory /srv/www/shop.gassycat.be/htdocs/img>
                AllowOverride All
                Require all granted
        </Directory>
        
        <Directory /srv/www/shop.gassycat.be/htdocs/themes>
                AllowOverride All
                Require all granted
        </Directory>
        
        <Directory /srv/www/shop.gassycat.be/htdocs/modules>
                AllowOverride All
                Require all granted
        </Directory>
```

#### 2. Set up _admin vhost_

Just a normal vhost point to Prestashop, but on the _admin vhost_, the first location restriction is the same but adjusted slightly to exclude `admin` from the restriction, of course:

```apacheconf
        <LocationMatch "^/(app|bin|classes|config|controllers|docs|translations|vendor)">
                Require all denied
        </LocationMatch>
```

#### 3. Add _admin vhost_ as a valid hostname in PrestaShop

The _admin vhost_ hostname is added to shop ID `1` as an alias (but not set as the main hostname), so that PS doesn't want to redirect to the main shop. So in `ps_shop_url` we have something like this:

| id_shop_url | id_shop | domain | domain_ssl | physical_uri | virtual_uri | main | active |
|-|-|-|-|-|-|-|-|
| 1 | 1 | shop.example.com | shop.example.com | / | | 1 | 1 |
| _2_ | _1_ | _admin.shop.example.com_ | _admin.shop.example.com_ | _/_ | | _0_ | _1_ |
| 3 ... | ... | ... | ... | ... | ... | ... | ... |

#### 4. Disable `PS_CANONICAL_REDIRECT`

Maybe unnecessary, but we have disabled `PS_CANONICAL_REDIRECT` for shop ID `1` too. Therefore, in `ps_configuration` it's like this:

| id_configuration | id_shop_group | id_shop | name | value | date_add | date_upd |
|-|-|-|-|-|-|-|
| 1000 | [NULL] | [NULL] | PS_CANONICAL_REDIRECT | 1 | 0000-00-00 00:00:00 | 0000-00-00 00:00:00 |
| _1001_ | _[NULL]_ | _1_ | _PS_CANONICAL_REDIRECT_ | _0_ | _0000-00-00 00:00:00_ | _0000-00-00 00:00:00_ |
| ... | ... | ... | ... | ... | ... | ... |

#### 5. Override `Link::getAdminBaseLink()`

Finally, we have to override `Link::getAdminBaseLink()` to make sure the back-office links are correct. This done by adding/modifying the file `/override/classes/Link.php` to be like this:

```php
<?php

class Link extends LinkCore
{
    /**
     * Patch from Ashley Butcher:
     * 
     * Always use the dedicated Back Office hostname.
     *
     * The public shop URL is deliberately unrelated to the Back Office URL.
     * We therefore don't use PrestaShop's multistore shop URL resolution here.
     *
     * @param int|null $idShop
     * @param bool|null $ssl
     * @param bool $relativeProtocol
     *
     * @return string
     */
    public function getAdminBaseLink($idShop = null, $ssl = null, $relativeProtocol = false)
    {
        if ($relativeProtocol) 
        {
            return '//admin.shop.example.com/';
        }
	
        return 'https://admin.shop.example.com/';
    }
}
```

---

We have only had one module misbehave with this change in place, and that's because that module was generating their own URLs and not following the API. It was straightforward to fix. Beyond that, everything else has been working well for a long while now.

Maybe this information is useful for someone else, or is a good starting point for someone who discovered this discussion and wants to lock down their back office.
